webapp/app.py

```diff
@@ -26,13 +31,30 @@ class User(UserMixin):
     about logged in users.
     """
 
-    def __init__(
-        self, id: str, name: str, email: str | None = None  # noqa: A002
-    ) -> None:
-        self.id = id
+    def __init__(self, mit_id: str, name: str, email: str | None = None) -> None:
+        self.mit_id = mit_id
         self.name = name
         self.email = email
 
+    def get_id(self) -> str:
+        try:
+            return str(self.mit_id)
+        except AttributeError:
+            raise NotImplementedError(
+                "No `mit_id` attribute - override `get_id`"  # noqa: EM101
+            ) from None
```

Comment:
Liking the direction of this (as discussed)!

Sorry to be nit-picky (again, seeing as this is our first implementation of this pattern, just trying to be thorough), but I'm feeling like maybe we don't need the `NotImplementedError`.  I think that was there primarily to document that it can/should be overridden from the default implementation.

I'd vote for very simple, letting errors bubble up if we encounter:

```python
def get_id(self) -> str:
    return str(self.mit_id)
```

Reply:
Ah, that is a valid and good point. I've simplified the `get_id` method as you suggested.